tools/hls-verifier/src/HlsVhdlTb.cpp

```diff
@@ -626,188 +531,127 @@ string HlsVhdlTb::getMemoryInstanceGeneration() {
-      out << "	file_open(fstatus, fp, OUTPUT_" << cDuvParam.parameterName
-          << ", WRITE_MODE);\n";
-      out << "	if (fstatus /= OPEN_OK) then\n";
-      out << "		assert false report \"Open file \" & OUTPUT_"
-          << cDuvParam.parameterName << " & \" failed!!!\" severity note;\n";
-      out << "		assert false report \"ERROR: Simulation using HLS TB "
-             "failed.\" severity failure;\n";
-      out << "	end if;\n";
-      out << "	write(token_line, string'(\"[[[runtime]]]\"));\n";
-      out << "	writeline(fp, token_line);\n";
-      out << "	file_close(fp);\n";
-      out << "	while transaction_idx /= TRANSACTION_NUM loop\n";
-      out << "		wait until tb_clk'event and tb_clk = '1';\n";
-      out << "	end loop;\n";
-      out << "	wait until tb_clk'event and tb_clk = '1';\n";
-      out << "	wait until tb_clk'event and tb_clk = '1';\n";
-      out << "	file_open(fstatus, fp, OUTPUT_" << cDuvParam.parameterName
-          << ", APPEND_MODE);\n";
-      out << "	if (fstatus /= OPEN_OK) then\n";
-      out << "		assert false report \"Open file \" & OUTPUT_"
-          << cDuvParam.parameterName << " & \" failed!!!\" severity note;\n";
-      out << "		assert false report \"ERROR: Simulation using HLS TB "
-             "failed.\" severity failure;\n";
-      out << "	end if;\n";
-      out << "	write(token_line, string'(\"[[[/runtime]]]\"));\n";
-      out << "	writeline(fp, token_line);\n";
-      out << "	file_close(fp);\n";
-      out << "	wait;\n";
-      out << "end process;\n";
+
+      os << llvm::formatv(procWriteTransactions, cDuvParam.parameterName,
+                          cDuvParam.parameterName, cDuvParam.parameterName,
+                          cDuvParam.parameterName, cDuvParam.parameterName);
     }
   }
-  out << "---------------------------------------------------------------------"
-         "-------\n\n";
-  return out.str();
 }
 
-string HlsVhdlTb::generateVhdlTestbench() {
-  stringstream tbOut;
-  tbOut << getLibraryHeader() << endl;
-  tbOut << getEntitiyDeclaration() << endl;
-  tbOut << getArchitectureBegin() << endl;
-  tbOut << getDuvInstanceGeneration() << endl;
-  tbOut << getMemoryInstanceGeneration() << endl;
-  tbOut << getOutputTagGeneration() << endl;
-  tbOut << getCommonBody() << endl;
-  tbOut << getArchitectureEnd() << endl;
-  return tbOut.str();
+void HlsVhdlTb::generateVhdlTestbench(mlir::raw_indented_ostream &os) {
+  os << VHDL_LIBRARY_HEADER;
+  os << "entity " + tleName + " is\n";
+  os << "end entity " + tleName + ";\n\n";
+  os << "architecture behavior of " << tleName << " is\n\n";
+  os.indent();
+  getConstantDeclaration(os);
+  getSignalDeclaration(os);
+  os.unindent();
+  os << "begin\n\n";
+  os.indent();
+  getDuvInstanceGeneration(os);
+  getMemoryInstanceGeneration(os);
+  getOutputTagGeneration(os);
+  os << COMMON_TB_BODY;
+  os.unindent();
+  os << "end architecture behavior;\n";
 }
 
 int HlsVhdlTb::getTransactionNumberFromInput() {
```

Comment:
This function is not related to your PR but I think line 666 is strange? I think it should always return `1` and we can remove the `hasInput` variable?